### ISSUE TITLE
Possible bug with Smart-Smiles translate Atom that without hydrogen

### ISSUE THREAD

**Issue by jingxingzhi** [rdkit/rdkit#4159]
**Describe the bug**
When I try to translate Atom without hydrogen between Smart-Smiles, I find RDkit will change the molecular Structure.

**To Reproduce**
When the atom is not fully bonded and connected to more than one hydrogen， it works well:
```
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#7H1]')))#get C[NH]
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#6H1]')))#get[CH]C
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#6H2]')))#get[CH2]C
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#8H1]')))#get[OH]
print(Chem.MolToSmarts(Chem.MolFromSmiles( 'C[NH]')))#get[#6]-[#7H]
print(Chem.MolToSmarts(Chem.MolFromSmiles( '[OH]')))#get[#8H]
```
![1621480441(1)](https://user-images.githubusercontent.com/37572298/118913847-cea54700-b95c-11eb-9d7e-75b05c29cf7b.jpg)
When the atom is not fully bonded and not connected to any hydrogen， it works wrong:
```
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#7H0]'))) #get CN
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#6H0]'))) #get CC
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#6H0]'),allHsExplicit=True) ) #get [CH3][CH3]
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#8H0]'))) #get O
print(Chem.MolToSmarts(Chem.MolFromSmiles( 'C[N]'))) #get [#6]-[#7]
print(Chem.MolToSmarts(Chem.MolFromSmiles( '[O]'))) #get [#8]
print(Chem.MolToSmarts(Chem.MolFromSmiles( '[C]'))) #get [#6]
```
![1621480493(1)](https://user-images.githubusercontent.com/37572298/118913849-cfd67400-b95c-11eb-84de-c5cdb6c8f8e8.jpg)
this code also not works:
```print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#8H0]'))) #get CO 
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#8&H0]')))#get CO 
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#8!H]')))#get CO 
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#8&!H]')))#get CO 
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#8X1]')))#get CO 
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#8&X1]')))#get CO 
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#8v1]')))#get CO 
print(Chem.MolToSmiles(Chem.MolFromSmarts( '[#6]-[#8&v1]')))#get CO
```
![1621480530(1)](https://user-images.githubusercontent.com/37572298/118913841-ccdb8380-b95c-11eb-9e5e-b2a11ecd95b0.jpg)
**Expected behavior**
I want translate some smarts wo smiles then get C[N],C[O],C[C]

**Configuration (please complete the following information):**
 - RDKit version: 2021.03.2
 - OS: win10 20H2
 - Python version (if relevant): 3.8
 - Are you using conda? yes
 - If you are using conda, which channel did you install the rdkit from? conda

**Additional context**
Well,match part can work well
```
m = Chem.MolFromSmiles('OCC[O]')
print(m.GetSubstructMatches(Chem.MolFromSmarts('[O&!H]')))#get ((3,),)
Draw.MolToImage(m,highlightAtoms=m.GetSubstructMatches(Chem.MolFromSmarts('[$([OH0])]'))[-1])
```
![1621480554(1)](https://user-images.githubusercontent.com/37572298/118913863-d664eb80-b95c-11eb-8413-eaf5f9929581.jpg)

**Comment by greglandrum** [comment]
@jingxingzhi : It's not possible to convert a molecule constructed from SMARTS into SMILES in a lossless way. We do the best we can, but the SMARTS parser adds query information to the molecule which doesn't get translated into the SMILES output.
Many, many SMARTS, like `C[N,O,S]`, can't be translated into SMILES at all.

We might be able to help more and provide alternate ideas if you describe why you want to do this.

**Comment by jingxingzhi** [comment]
> @jingxingzhi：不可能将SMARTS构造的分子无损地转化为SMILES。我们会尽力而为，但是SMARTS解析器会将查询信息添加到未翻译为SMILES输出的分子中。
> 像一样的许多许多SMARTS`C[N,O,S]`根本无法翻译成SMILES。
> 
> 如果您描述为什么要这样做，我们也许可以提供更多帮助并提供其他想法。

Thanks a lot.
actualy 
I was try to write an reaction and it will creats atom which is not fully bonded(Free radicals)
just like COC→[CH3]+C[O]
but when i write:
```
AllChem.ReactionFromSmarts('[#6]-[#8:1]-[#6:2]>>[#6H3].[#6]-[#8X0]')
```
it not works as I expected
```
rxn = AllChem.ReactionFromSmarts('[#6]-[#8:1]-[#6:2]>>[#6H3].[#6]-[#8X0]')
abc=Chem.MolFromSmiles('COC')
reactants = (abc,)
res=rxn.RunReactants(reactants)
try:
    Draw.MolsToGridImage([abc]+list(res[0]))
except:
    pass
Draw.MolsToGridImage([abc])
print(res[0][0])
Draw.MolsToGridImage([Chem.MolFromSmiles(Chem.MolToSmiles(res[0][0])),Chem.MolFromSmiles(Chem.MolToSmiles(res[0][1]))])
```
I can not get C[O]
![1621482322(1)](https://user-images.githubusercontent.com/37572298/118916194-06ae8900-b961-11eb-9f91-364f05820c89.jpg)

if I write：
```
Draw.MolsToGridImage([res[0][0],res[0][1]])
```
it will show
```
RuntimeError                              Traceback (most recent call last)
<ipython-input-162-0a73e9137d75> in <module>
----> 1 Draw.MolsToGridImage([res[0][0],res[0][1]])

C:\ProgramData\Anaconda3\lib\site-packages\rdkit\Chem\Draw\IPythonConsole.py in ShowMols(mols, maxMols, **kwargs)
    195       if prop in kwargs:
    196         kwargs[prop] = kwargs[prop][:maxMols]
--> 197   res = fn(mols, drawOptions=drawOptions, **kwargs)
    198   if kwargs['useSVG']:
    199     return SVG(res)

C:\ProgramData\Anaconda3\lib\site-packages\rdkit\Chem\Draw\__init__.py in MolsToGridImage(mols, molsPerRow, subImgSize, legends, highlightAtomLists, highlightBondLists, useSVG, returnPNG, **kwargs)
    609                           highlightBondLists=highlightBondLists, **kwargs)
    610   else:
--> 611     return _MolsToGridImage(mols, molsPerRow=molsPerRow, subImgSize=subImgSize, legends=legends,
    612                             highlightAtomLists=highlightAtomLists,
    613                             highlightBondLists=highlightBondLists, returnPNG=returnPNG, **kwargs)

C:\ProgramData\Anaconda3\lib\site-packages\rdkit\Chem\Draw\__init__.py in _MolsToGridImage(mols, molsPerRow, subImgSize, legends, highlightAtomLists, highlightBondLists, drawOptions, returnPNG, **kwargs)
    552           setattr(dops, k, v)
    553           del kwargs[k]
--> 554     d2d.DrawMolecules(list(mols), legends=legends or None, highlightAtoms=highlightAtomLists,
    555                       highlightBonds=highlightBondLists, **kwargs)
    556     d2d.FinishDrawing()

RuntimeError: Pre-condition Violation
	getExplicitValence() called without call to calcExplicitValence()
	Violation occurred on line 199 in file Code\GraphMol\Atom.cpp
	Failed Expression: d_explicitValence > -1
	RDKIT: 2021.03.2
	BOOST: 1_74

```

**Comment by greglandrum** [comment]
@jingxingzhi you have to map the atoms in the reaction if you want things to work correctly.
You also should always sanitize the results which come back from reactions (this is mentioned in the docs: https://rdkit.org/docs/GettingStartedInPython.html#chemical-reactions)
Here's an example:
```
In [6]: rxn = AllChem.ReactionFromSmarts('[#6:3]-[#8:1]-[#6:2]>>[#6H3:3].[#6:2]-[#8H0:1]')

In [7]: ps = rxn.RunReactants((Chem.MolFromSmiles('COC'),))

In [9]: for p in ps: Chem.SanitizeMol(p[0])

In [10]: [Chem.MolToSmiles(p[0]) for p in ps]
Out[10]: ['[CH3]', '[CH3]']

In [11]: for p in ps: Chem.SanitizeMol(p[1])

In [12]: [Chem.MolToSmiles(p[1]) for p in ps]
Out[12]: ['C[O]', 'C[O]']
```

Notice that the output molecules have the radicals you were looking for.

**Comment by jingxingzhi** [comment]
> @jingxingzhi you have to map the atoms in the reaction if you want things to work correctly.
> You also should always sanitize the results which come back from reactions (this is mentioned in the docs: https://rdkit.org/docs/GettingStartedInPython.html#chemical-reactions)
> Here's an example:
> 
> ```
> In [6]: rxn = AllChem.ReactionFromSmarts('[#6:3]-[#8:1]-[#6:2]>>[#6H3:3].[#6:2]-[#8H0:1]')
> 
> In [7]: ps = rxn.RunReactants((Chem.MolFromSmiles('COC'),))
> 
> In [9]: for p in ps: Chem.SanitizeMol(p[0])
> 
> In [10]: [Chem.MolToSmiles(p[0]) for p in ps]
> Out[10]: ['[CH3]', '[CH3]']
> 
> In [11]: for p in ps: Chem.SanitizeMol(p[1])
> 
> In [12]: [Chem.MolToSmiles(p[1]) for p in ps]
> Out[12]: ['C[O]', 'C[O]']
> ```
> 
> Notice that the output molecules have the radicals you were looking for.

Thats really help,I truly appreciate your support during this work. 

I'm just curious why the previous conversion method didn't work，so I read the SmilesWrite.cpp and the SmartsWrite.cpp. 
I find in SmilesWrite.cpp there was "needsBracket", while SmartsWrite.cpp does not has that. 
If the needsBracket Atom was incorrectly labeled as "AtomHasImplicitH"or"AtomHCount"  ?
'''
if (needsBracket) {    res += "[";  }
  if (isotope &amp;&amp; (isomericSmiles || (atom-&gt;hasOwningMol() &amp;&amp;                                     atom-&gt;getOwningMol().hasProp(                                         common_properties::_doIsoSmiles)))) {    res += std::to_string(isotope);  }  // this was originally only done for the organic subset,  // applying it to other atom-types is a fix for Issue 3152751:  if (!doKekule &amp;&amp; atom-&gt;getIsAromatic() &amp;&amp; symb[0] &gt;= 'A' &amp;&amp; symb[0] &lt;= 'Z') {    symb[0] -= ('A' - 'a');  }  res += symb;
  res += atString;
  if (needsBracket) {    unsigned int totNumHs = atom-&gt;getTotalNumHs();
'''

`Chem.MolFromSmiles(Chem.MolToSmiles(Chem.MolFromSmarts('[#8H1]')))` 
![1621482375(1)](https://user-images.githubusercontent.com/37572298/118953293-442a0b00-b98f-11eb-968f-9cbc81aa7b4c.jpg)

`Chem.MolFromSmiles(Chem.MolToSmiles(Chem.MolFromSmarts('[#6]-[#8H0]')))`
![1621502236(1)](https://user-images.githubusercontent.com/37572298/118953309-47bd9200-b98f-11eb-840b-72fde1948e53.jpg)

You can see this issuewas only appeared when Atom radicals has 0 H,if there is a better way to deal with this problem？
I think may support he mutual transformation of A and B can be help ,just like find Atom that "needsBracket" , re-count thenum of H and accept [#numH0]?

thank you for your help.
